run/service-auth/app.py

```diff
@@ -12,30 +12,133 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# [START auth_validate_and_decode_bearer_token_on_flask]
+# [START cloudrun_service_to_service_receive]
+"""Demonstrates how to receive authenticated service-to-service requests
+on a Cloud Run Service.
+"""
+
 from http import HTTPStatus
 import os
+from typing import Optional
+from urllib.request import Request, urlopen
 
 from flask import Flask, request
 
-from receive import receive_request_and_parse_auth_header
+from google.auth.exceptions import GoogleAuthError
+from google.auth.transport import requests
+from google.cloud import run_v2
+from google.oauth2 import id_token
+
+# Get the Service Name as found in Cloud Run.
+SERVICE_NAME = os.getenv("K_SERVICE")
+
+# Get the Project ID.
+req = Request("http://metadata.google.internal/computeMetadata/v1/project/project-id")
+req.add_header("Metadata-Flavor", "Google")
+PROJECT_ID = urlopen(req).read().decode("utf-8")
+
+# Get the Region.
+req = Request("http://metadata.google.internal/computeMetadata/v1/instance/region")
+req.add_header("Metadata-Flavor", "Google")
+
+# Returns "projects/PROJECT-NUMBER/regions/REGION"
+project_region_list = urlopen(req).read().decode("utf-8").split('/')
+
+REGION = project_region_list[3]
+
+# Get the Service URL, required to define the valid audience for the Token.
+# https://cloud.google.com/run/docs/triggering/https-request#deterministic
+FULL_SERVICE_NAME = f"projects/{PROJECT_ID}/locations/{REGION}/services/{SERVICE_NAME}"
+
+client = run_v2.ServicesClient()
+
+service_request = run_v2.GetServiceRequest(
+    name=FULL_SERVICE_NAME,
+)
+
+SERVICE_URI = client.get_service(request=service_request).uri
 
 app = Flask(__name__)
 
 
+def parse_auth_header(auth_header: str) -> Optional[str]:
+    """Parse the authorization header, validate and decode the Bearer token.
+
+    Args:
+        auth_header: Raw HTTP header with a Bearer token.
+
+    Returns:
+        A string containing the email from the token.
+        None if the token is invalid or the email can't be retrieved.
+    """
+
+    # Split the auth type and value from the header.
+    try:
+        auth_type, creds = auth_header.split(" ", 1)
+    except ValueError:
+        print("Malformed Authorization header.")
+        return None
+
+    # The token audience will be the SERVICE_URL.
+    # If `audience` was None, it won't be verified.
+    audience = SERVICE_URI
```

Comment:
This comment seems to imply this audience shouldn't be forced to be set here?
This should be clarified.